From f0f5a49aee8647539eafb046b629532d14c551da Mon Sep 17 00:00:00 2001
From: Lars-Dominik Braun <lars@6xq.net>
Date: Mon, 15 Feb 2021 14:06:25 +0100
Subject: [PATCH] rserver: Add option to terminate server after session
 terminates
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a oneshot option, which is useful in combination with
`--auth-none=1` and rsession’s `--session-timeout-minutes`. Terminating
the session will also quit the server, so it won’t linger if unused.
---
 src/cpp/server/ServerMain.cpp      | 6 +++++-
 src/cpp/server/server-options.json | 7 +++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/cpp/server/ServerMain.cpp b/src/cpp/server/ServerMain.cpp
index 12f9ceaaf6..ac0410127a 100644
--- a/src/cpp/server/ServerMain.cpp
+++ b/src/cpp/server/ServerMain.cpp
@@ -383,6 +383,9 @@ Error waitForSignals()
       if (sig == SIGCHLD)
       {
          sessionManager().notifySIGCHLD();
+         if (server::options().serverOneshot()) {
+            return Success();
+         }
       }
 
       // SIGINT, SIGQUIT, SIGTERM
@@ -795,7 +798,8 @@ int main(int argc, char * const argv[])
       if (error)
          return core::system::exitFailure(error, ERROR_LOCATION);
 
-      // NOTE: we never get here because waitForSignals waits forever
+      s_pHttpServer->stop();
+
       return EXIT_SUCCESS;
    }
    CATCH_UNEXPECTED_EXCEPTION
diff --git a/src/cpp/server/server-options.json b/src/cpp/server/server-options.json
index f1dc1c22b8..38af55d299 100644
--- a/src/cpp/server/server-options.json
+++ b/src/cpp/server/server-options.json
@@ -50,6 +50,13 @@
             "defaultValue": {"code": "core::system::effectiveUserIsRoot()", "description": "1 (true) if rserver was run with root privilege, otherwise 0 (false)."},
             "description": "Indicates whether or not the rserver process should run as a daemon."
          },
+         {
+            "name": "server-oneshot",
+            "memberName": "serverOneshot_",
+            "type": "bool",
+            "defaultValue": false,
+            "description": "Terminate server when rsession terminates."
+         },
          {
             "name": "server-pid-file",
             "memberName": "serverPidFile_",
-- 
2.26.2

